### PR TITLE
adding the YOLACT instance segmentation model

### PR DESCRIPTION
**Pull request by Abdullah-Elkasaby**

### CODE REVIEW

**Comment by adrianboguszewski** [comment]
Thank you for your contribution. Before we do the review please do the self-check first (copy the list and tick the item when it's done):
- [ ] each function is described by docstrings and type hints
- [ ] notebook contains explicit descriptions and explanatory diagrams
- [ ] the notebook doesn't use any data (image, video, etc.) that is not CC4.0 licensed 
- [ ] there is a README.md file in consistent style (look at other notebooks)
- [ ] the notebook is added to the main README
- [ ] there are no grammar, punctuation or typo issues (use any free tool for that e.g. Grammarly)
- [ ] there are no committed files besides notebook and readme (please use images or videos from data dir)
- [ ] your PR doesn't change any other notebooks
- [ ] all CI checks passed

**Comment by adrianboguszewski** [comment]
And it shouldn't be as submodule but committed directly :)

**Comment by Abdullah-Elkasaby** [comment]
@adrianboguszewski  Thanks for the feedback,  I will be working on the changes.

**Comment by Abdullah-Elkasaby** [comment]
@adrianboguszewski 
I have made some changes and ran the treon CI checks and passed, could you have a look?

- [x] Each function is described by docstrings and type hints.
- [x] The notebook contains explicit descriptions and explanatory diagrams.
- [x] The notebook doesn't use any data (image, video, etc.) that is not CC4.0 licensed.
- [x] There is a README.md file in consistent style (look at other notebooks).
- [x] The notebook is added to the main README.
- [x] There are no grammar, punctuation, or typo issues (use any free tool for that, e.g., Grammarly).
- [x] There are no committed files besides the notebook and README (please use images or videos from the data directory).
- [x] Your PR doesn't change any other notebooks.
- [x] Treon CI checks passed.

**Comment by adrianboguszewski** [comment]
Two more comments:
- What is the utils directory? Do you need all of these utils?
- Descriptions are missing. The notebook is educational content, so it should contain an explanation of why it was done like that. See other notebooks for the reference

**Comment by Abdullah-Elkasaby** [comment]
> Two more comments:
> 
> * What is the utils directory? Do you need all of these utils?
> * Descriptions are missing. The notebook is educational content, so it should contain an explanation of why it was done like that. See other notebooks for the reference

The model does need all of those, the original model was even more complex and had multiple dependencies to multiple directories.
Are a description in the readme file and  comments in the notebook of the utils directory enough?

**Comment by Abdullah-Elkasaby** [comment]
@adrianboguszewski I have some changes and the code check passed, however there's some error regarding Binder that I don't get.

**Comment by adrianboguszewski** [comment]
This output also doesn't look good (like wrong output - mixed areas)
image

**Comment by Abdullah-Elkasaby** [comment]
@adrianboguszewski  the drive is not mine, it's available on the the main repo of the model, I can set it up on my drive for now, however I don't know if that will be suitable. as for the mixed areas problem, I will try to to work on but I might need some help.

**Comment by adrianboguszewski** [comment]
If it's pointing to the main repo, please find a way to skip that access deny error

**Comment by adrianboguszewski** [comment]
@Abdullah-Elkasaby, any plans to continue working on this?

**Comment by Abdullah-Elkasaby** [comment]
> any plans to continue working on this?

Yes, I do have plans to continue working on this project. However, please note that there might be some delays or gaps in my availability from time to time.

**Comment by adrianboguszewski** [comment]
Ok. So please look at the above comments and fix them first.

**Comment by Abdullah-Elkasaby** [comment]
> Ok. So please look at the above comments and fix them first.

I think I have fixed most of them, I will be working on the output problem.

**Comment by Abdullah-Elkasaby** [comment]
@adrianboguszewski  could you  have a look at the output now ? 
 also the failing checks are related to renaming the PR.

**Comment by Abdullah-Elkasaby** [comment]
@adrianboguszewski the failing code check is not related to my notebook, could you please check that?

**Comment by adrianboguszewski** [comment]
@Abdullah-Elkasaby, thanks for the changes. It looks better. I like that notebook :) 

However, please still consider the following points:
- Descriptions are still missing. I mean the descriptions between code cells. It would be best if you described above any cell what you're going to do
- You should use the org model drive, not yours. If it causes any problems, they have to be fixed first. You shouldn't use your own drive - it's a security vulnerability
- Don't worry about the PR labeler. It still doesn't work as expected
- Please remove unnecessary blank lines (those at the top and bottom of cells)
- Please remove all commented code. It should be either runnable (uncommented) or gone

**Comment by Abdullah-Elkasaby** [comment]
> @Abdullah-Elkasaby, thanks for the changes. It looks better. I like that notebook :)
> 
> However, please still consider the following points:
> 
> * Descriptions are still missing. I mean the descriptions between code cells. It would be best if you described above any cell what you're going to do
> * You should use the org model drive, not yours. If it causes any problems, they have to be fixed first. You shouldn't use your own drive - it's a security vulnerability
> * Don't worry about the PR labeler. It still doesn't work as expected
> * Please remove unnecessary blank lines (those at the top and bottom of cells)
> * Please remove all commented code. It should be either runnable (uncommented) or gone

I have made these changes, but the org drive was causing a problem for you locally which I don't know what is it about

**Comment by adrianboguszewski** [comment]
Ok. Two more things to do:
- there are still blank lines at the top or bottom of some code cells - please find them and remove
- thanks for the descriptions - please convert them from headings to paragraphs. Headings should separate bigger sections of related cells only

**Comment by Abdullah-Elkasaby** [comment]
@adrianboguszewski I have add these changes, but unfortunately the docker test is now failing idk why.

**Comment by Abdullah-Elkasaby** [comment]
> Don't worry about anything not related to your changes.
> 
> * The gitignore file doesn't seem to be needed in this PR, does it?
> * We're avoiding committing new image files. The image with results should be an outcome from the code, not loaded from disk.

the image output is notebook is not from disk, it's coming from the model, however that image on the disk is used as a thumbnail in the readmefile, I will delete it and use the image from a seperate repo.

I used the gitignore file to keep the downloaded model and the weights local as they are downloaded from google drive, I can delete it though

**Comment by adrianboguszewski** [comment]
> > Don't worry about anything not related to your changes.
> > 
> > * The gitignore file doesn't seem to be needed in this PR, does it?
> > * We're avoiding committing new image files. The image with results should be an outcome from the code, not loaded from disk.
> 
> the image output is notebook is not from disk, it's coming from the model, however that image on the disk is used as a thumbnail in the readmefile, I will delete it and use the image from a seperate repo.

In README files we use the links to images uploaded to GitHub. See the contribution guide.

> 
> I used the gitignore file to keep the downloaded model and the weights local as they are downloaded from google drive, I can delete it though

Gitignore is not needed to ignore those files. You can just skip them when committing. 

It looks good to me. Thanks! I'm approving it and passing it to the next reviewer :)

**Review by adrianboguszewski — APPROVED** [review]

**Comment by eaidova** [comment]
@Abdullah-Elkasaby generally, notebook is good, but do you really need so many files and manual model construction inside notebook? Can we just clone model repo and install it instead?

**Comment by eaidova** [comment]
please also rename notebook to keep the same name like notebook subfolder

**Comment by Abdullah-Elkasaby** [comment]
> @Abdullah-Elkasaby generally, notebook is good, but do you really need so many files and manual model construction inside notebook? Can we just clone model repo and install it instead?

The original model uses more dependencies to many different directories and files, I tried to to keep a bit organized into one directory and keep the logic related to building the model in the nootebook. 
As for the comments, alot of them are copied from the original model's repo.
I renamed the notebook and the directory to match and other changes too 
@eaidova

**Comment by eaidova** [comment]
> > generally, notebook is good, but do you really need so many files and manual model construction inside notebook? Can we just clone model repo and install it instead?
> 
> The original model uses more dependencies to many different directories and files, I tried to to keep a bit organized into one directory and keep the logic related to building the model in the nootebook. As for the comments, alot of them are copied from the original model's repo. I renamed the notebook and the directory to match and made some other minor changes too. @eaidova

@Abdullah-Elkasaby I think there should be a balance between independency and flexibility. It can be difficult to map original model code and tutorial in case, when user want to apply it on own model e.g. pretrained on different dataset or which has different backbone (the steps will be similar, right?), that is why I think using original code structure will have more value than replication of files in notebook and building model from scratch.

you can clone repo inside notebook using command:
!git clone <repo>
then you can install it or add in sys.path or move inside repo using %cd to make code visible in notebook python environment.

then use original imports and build model in notebook.

I think it should simplify notebook code.

**Comment by Abdullah-Elkasaby** [comment]
> then use original imports and build model in notebook.

The original model does not support conversion out of the box, there has to be changes made to it to support that as mentioned [HERE](https://github.com/dbolya/yolact/issues/74#issuecomment-504762356) . So, I don't think that will work that easily without dealing with those changes first.

I have also tested chaning the backbone for the model and it  works fine without needing to change anything.
image

**Comment by eaidova** [comment]
> > then use original imports and build model in notebook.
> 
> The original model does not support conversion out of the box, there has to be changes made to it to support that as mentioned [HERE](https://github.com/dbolya/yolact/issues/74#issuecomment-504762356) . So, I don't think that will work that easily without dealing with those changes first.
> 
> I have also tested chaning the backbone for the model and it works fine without needing to change anything. image

I understand that it works, I think it is not easy to make it decision because code rewriting if people slightly know original without a lot of details. 
Possibly this tutorial should also can be helpful for export model 
https://docs.openvino.ai/2023./openvino_docs_MO_DG_prepare_model_convert_model_pytorch_specific_Convert_YOLACT.html

**Comment by Abdullah-Elkasaby** [comment]
> I understand that it works, I think it is not easy to make it decision because code rewriting if people slightly know original without a lot of details. Possibly this tutorial should also can be helpful for export model https://docs.openvino.ai/2023./openvino_docs_MO_DG_prepare_model_convert_model_pytorch_specific_Convert_YOLACT.html

I have tested that before but the patch didn't work as intended and I have just tested it again, still does not work 

also could you clarify the problem a bit and what should be done ?

**Comment by eaidova** [comment]
> > I understand that it works, I think it is not easy to make it decision because code rewriting if people slightly know original without a lot of details. Possibly this tutorial should also can be helpful for export model https://docs.openvino.ai/2023./openvino_docs_MO_DG_prepare_model_convert_model_pytorch_specific_Convert_YOLACT.html
> 
> I have tested that before but the patch didn't work as intended and I have just tested it again, still does not work
> 
> also could you clarify the problem a bit and what should be done ?

@Abdullah-Elkasaby my concern that in such amount of code copied on notebook side may lead of loosing focus on educational steps required to enable model in openvino (it is the main goal of our repo - not only speak about models, but provide instructions how to convert and run model using openvino), can you please to think how to better represent it? maybe clean up unnecessary thing, provide some explanation which places was modified from original code. The main idea to provide user instruction how to make deal with yolact model using openvino, notebook should be easy to adopt to other models in this family and it is should be clear that some steps can be reused

**Comment by Abdullah-Elkasaby** [comment]
> 

Does making the notebook code only related  to the conversion and use in openvino achieve this? 
My approach would be moving the code related to initilazing the mode to the same files that were used in the original model. 

The changes made to the original repo are similar to the ones in the previously mentioned Patch, I just did them manually and can mention them if that will be better.

**Comment by Abdullah-Elkasaby** [comment]
@eaidova  
could you check the notebook now, also mentioining the changes will be hard to do and will require doing things from the begining, but they are commented out in each file if anyone wanted to replicate it on their own

**Comment by Abdullah-Elkasaby** [comment]
@eaidova 
could you have a look at the notebook ?

**Comment by Abdullah-Elkasaby** [comment]
@eaidova @dmatveev @kuujo  any updates?

**Comment by dmatveev** [comment]
Sorry I don't have any updates, sorry

**Comment by adrianboguszewski** [comment]
I'm closing this due to inactivity for a long time. Nevertheless, thank you for contributing and applying for Google Summer of Code. We encourage you to consider GSoC 2024 as well :)

**Comment by Abdullah-Elkasaby** [comment]
@adrianboguszewski I just noticed the closing of my PR :(
I was not inactive, I just didn't have any reviews from your side for a long time, look at my latest commits, they are just renaming the directory number because another module has been accepted with the same number so I had to rename it, mine wasn't even looked at

Any chance of being opened again and then reviewed ?

File: notebooks/274-yolact-instance-segmentation-ov-demo/yolact-instance-segmentation-ov-demo.ipynb
Comment (aleksandr-mokrov) [inline notebooks/274-yolact-instance-segmentation-ov-demo/yolact-instance-segmentation-ov-demo.ipynb:1]: Instead `from openvino.tools.mo import convert_model` use `import openvino as ov` and `ov_model = ov.convert_model(net, example_input=batch)`. And here you don't need to use onnx, openvino can handle it.
Instead `ie = Core()` use `core = ov.Core()`
Also it is a good practice to check if model is already converted and skip this step, like this:
```
ov_model_xml_path = Path('models/ov_model.xml')

if not ov_model_xml_path.exists():
    ov_model = ov.convert_model(
        model,
        example_input=dummy_input
    )
    ov.save_model(ov_model, ov_model_xml_path)
```

Comment (aleksandr-mokrov) [inline notebooks/274-yolact-instance-segmentation-ov-demo/yolact-instance-segmentation-ov-demo.ipynb:1]: Please add an interactive example with gradio.

**Review by aleksandr-mokrov — CHANGES_REQUESTED** [review]

File: notebooks/274-yolact-instance-segmentation-ov-demo/yolact-instance-segmentation-ov-demo.ipynb
[the remark this answers is not in the document]
Comment (Abdullah-Elkasaby) [reply notebooks/274-yolact-instance-segmentation-ov-demo/yolact-instance-segmentation-ov-demo.ipynb:1]: reploaced them

[the remark this answers is not in the document]
Comment (Abdullah-Elkasaby) [reply notebooks/274-yolact-instance-segmentation-ov-demo/yolact-instance-segmentation-ov-demo.ipynb:1]: changed the reference for the demo image, could have a look at it ?

Comment (Abdullah-Elkasaby) [reply notebooks/274-yolact-instance-segmentation-ov-demo/yolact-instance-segmentation-ov-demo.ipynb:1]: thank you for your feedback, this is indeed much easier

**Comment by Abdullah-Elkasaby** [comment]
@aleksandr-mokrov could you have a look at it now ?

**Comment by Abdullah-Elkasaby** [comment]
@aleksandr-mokrov @eaidova any updates?

File: notebooks/274-yolact-instance-segmentation-ov-demo/README.md [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+
+* Download the pretrained weights.
+* Prepare the PyTorch model.
+* Validate the original model.
+* Convert the PyTorch model to ONNX.
+* Convert the ONNX model to OpenVINO IR.
+* Validate the converted model.
+* Run the model 
+
```
[the remark this answers is not in the document]
Comment (aleksandr-mokrov) [reply notebooks/274-yolact-instance-segmentation-ov-demo/README.md]: The main idea here is that you should be able create new virtual environment, install jupyterlab and ipywidgets, and then run your notebook that will install all necessary requirements for your notebook.

**Comment by aleksandr-mokrov** [comment]
@Abdullah-Elkasaby currents PR affects a lot of files that are not relevant to current example. Looks like you merged main into you branch in the wrong way.

**Comment by Abdullah-Elkasaby** [comment]
> @Abdullah-Elkasaby currents PR affects a lot of files that are not relevant to current example. Looks like you merged main into you branch in the wrong way

I don't see the conficts on my side idk why

**Comment by aleksandr-mokrov** [comment]
> I don't see the conficts on my side idk why
There are no conflicts. Problem is that there changes that affects a lot of files from other notebooks. This happens when you do not resolve the conflict manually correctly. I believe that the problem in this commit: 
https://github.com/openvinotoolkit/openvino_notebooks/pull/1067/commits/ed491f31c042936d3aaf10db665d1a45d8f99f43

**Comment by Abdullah-Elkasaby** [comment]
> ed491f3

I have reverted back those changes, it should be up to date with the current repo now

**Comment by aleksandr-mokrov** [comment]
> > [ed491f3](https://github.com/openvinotoolkit/openvino_notebooks/commit/ed491f31c042936d3aaf10db665d1a45d8f99f43)
> 
> I have reverted back those changes, it should be up to date with the current repo now

Great! Check also files under .github/workflows diretory and .docker/Pipfile.lock. And resolve comments starting with https://app.reviewnb.com/openvinotoolkit/openvino_notebooks/pull/1067/discussion/#comment-1905093771. It will simplify this example.

**Comment by Abdullah-Elkasaby** [comment]
> > > [ed491f3](https://github.com/openvinotoolkit/openvino_notebooks/commit/ed491f31c042936d3aaf10db665d1a45d8f99f43)
> > 
> > 
> > I have reverted back those changes, it should be up to date with the current repo now
> 
> Great! Check also files under .github/workflows diretory and .docker/Pipfile.lock. And resolve comments starting with https://app.reviewnb.com/openvinotoolkit/openvino_notebooks/pull/1067/discussion/#comment-1905093771. It will simplify this example.

I have made the changes regarding making the code related to initilazing the model to the same  as the one  used in the original model.

However, I don't understand your comment about github workflows and docker directory, could you please clarify?

**Comment by aleksandr-mokrov** [comment]
> I have made the changes regarding making the code related to initilazing the model to the same as the one used in the original model.

There are a lot of copied files from yolact repository. Some of them has some changes. It is possible to avoid it. I left tips on how to do this. Such as cloning a repository and making changes to objects rather than files. I checked it and it works. There are few changes required to the notebook, but we eliminate the need to bring a bunch of third-party files into our repository.

> However, I don't understand your comment about github workflows and docker directory, could you please clarify?

Look into "File changed" tab. There are changes in github workflows and docker directory. These changes should not happen.

**Comment by aleksandr-mokrov** [comment]
<p>An example implementation of the last option to keep the same behavior but not change the source code:</p><p> </p><pre class="ql-syntax" spellcheck="false">def detect(pred_outs, net):
  return pred_outs['loc'], pred_outs['conf'], pred_outs['mask'], pred_outs['priors'], pred_outs['proto']

net.detect = detect
</pre>

---

 <a href='https://app.reviewnb.com/openvinotoolkit/openvino_notebooks/pull/1067/discussion/#comment-1905093847'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by andrei-kochin** [comment]
@Abdullah-Elkasaby any updates here on requested changes?

**Pull request closed without merging**